text/###-v2-provider-interface.md

```diff
@@ -0,0 +1,402 @@
+because we won't be performing needless conversions everytime we run
+experiments. Another general benefit is this will result in a cleaner and
+better documented interface for working with backends and results.
+
+Additionally, it makes the domains of what's part of IBMQ/IQX and what's part
+of terra much clearer. Right now to make changes to terra's interfaces this
+often gets caught up in how iqx works today and limits what is possible.
+
+## Design Proposal
+
+The transition will happen in 2 stages. The first stage, v1.5 which is
+a maintanence change for terra is to remove marshmallow from the existing
+providers interface. This will make things more transparent in how they are
+built simplifying andy time we need to work with the interface, but also
+significantly speed up the performance of the current interface. The first
+piece of this has been pushed for the qobj class here:
+
+https://github.com/Qiskit/qiskit-terra/pull/3383
+
+This basically just mirrors the marshmallow class heirarchy and replaces each
+class with a flat class with just attributes for each element in the schema.
+The removal of marshmallow will need to be replicated for backends, jobs, and
+results too before we can move on to the new interface.
+
+This is needed for 3 reasons. The first is that the providers interface has
+existed in it's current form for some time and is how any backend interfaces
+with qiskit. We need to keep this around not just for a smooth transition to
+the new interface but also for providers that are not maintained by our team.
+We want to maximize compatibility with any providers that exist out there and
+removing the current interface would potentially break a lot of people. The
+second is to simplify the maintanence around this existing interface. Right
+now the interface for interacting with providers is too obfuscated creating a
+flat object makes it much simpler to work with when we need to make additions
+or changes to keep the interface functioning. It also makes it easier for local
+simulators to use C aligned data structures natively because everything is
+just a normal python attribute and doesn't go through marshmallow's
+transformation layer. The third is that doing this will greatly improves
+performance. In Qiskit/qiskit-terra#3383 for assembling a qobj for 100
+experiments of circuits with 53 qubits and a depth of 2048 took < 2min with the
+PR removing marshmallow and 3min and 40 seconds with master (object generation
+is also significantly faster, and more importantly uses far less memory).
+
+This also starts the process of decoupling IQX's API from the provider interface
+because removing marshmallow means we don't get the 2 transformations it performs
+`numpy array -> list` and `complex numbers -> [re, im]`. Basically when v1.5
+starts merging we are saying that the `qobj.to_dict()`'s output is not the same
+as `json.loads()` from what the ibmq provider sends on the wire.
+
```

Comment:
I would add subsections for the different sections, v1.5 and v2.0

Reply:
Done in: https://github.com/Qiskit/rfcs/pull/2/commits/d854980f610eacb8c2fbcda2ff2cdf1be3464cbd